instrumentation/aws-sdk/aws-sdk-2.2/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v2_2/internal/BedrockRuntimeImpl.java

```diff
@@ -125,4 +210,91 @@ private static Double floatToDouble(Float value) {
     }
     return Double.valueOf(value);
   }
+
+  private static LogRecordBuilder newEvent(Context otelContext, Logger eventLogger) {
+    return eventLogger
+        .logRecordBuilder()
+        .setContext(otelContext)
+        .setAttribute(
+            GEN_AI_SYSTEM, BedrockRuntimeAttributesGetter.GenAiSystemIncubatingValues.AWS_BEDROCK);
+  }
+
+  private static void emitToolResultEvents(
+      Context otelContext, Logger eventLogger, Message message, boolean captureMessageContent) {
+    for (ContentBlock content : message.content()) {
+      if (content.toolResult() == null) {
+        continue;
+      }
+      Map<String, Value<?>> body = new HashMap<>();
+      body.put("id", Value.of(content.toolResult().toolUseId()));
+      if (captureMessageContent) {
+        StringBuilder text = new StringBuilder();
+        for (ToolResultContentBlock toolContent : content.toolResult().content()) {
+          if (toolContent.text() != null) {
+            text.append(toolContent.text());
+          }
+          if (toolContent.json() != null) {
+            text.append(serializeDocument(toolContent.json()));
+          }
+        }
+        body.put("content", Value.of(text.toString()));
+      }
+      newEvent(otelContext, eventLogger)
+          .setAttribute(EVENT_NAME, "gen_ai.tool.message")
+          .setBody(Value.of(body))
+          .emit();
+    }
+  }
+
+  private static Value<?> convertMessage(
+      Message message, int index, @Nullable StopReason stopReason, boolean captureMessageContent) {
+    StringBuilder text = null;
+    List<Value<?>> toolCalls = null;
+    for (ContentBlock content : message.content()) {
+      if (captureMessageContent && content.text() != null) {
+        if (text == null) {
+          text = new StringBuilder();
+        }
+        text.append(content.text());
+      }
+      if (content.toolUse() != null) {
+        if (toolCalls == null) {
+          toolCalls = new ArrayList<>();
+        }
+        toolCalls.add(convertToolCall(content.toolUse(), captureMessageContent));
+      }
+    }
+    Map<String, Value<?>> body = new HashMap<>();
```

Comment:
Ignore is fine as I expect it isn't normal, but for these simple structures I wonder if using Jackson objects are as simple as a map and would be less overhead. Probably not and also in context of everything else not important to optimize just because I looked.

Reply:
These are OTel `Value`, not JSON so either this or an array of `KeyValue` using `toArray`. I wish for something like `Value.mapBuilder` or something that could be more optimized, or at least a method accepting `Iterable`.

Reply:
Ah right i didn't look at where Value was imported gotcha